### PR TITLE
fix: support grad clipping for TP through replicating non-sharded modules

### LINKED ISSUE #36296: tensor parallel training bug
### System Info

transformers：4.45.dev0
python：3.11
linux

### Who can help?

#34194 

### Information

- [x] The official example scripts
- [ ] My own modified scripts

### Tasks

- [x] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

torchrun  --nnodes 1 --nproc_per_node 2 --master_port 27654 run_clm.py \
    --model_name_or_path TinyLlama/TinyLlama-1.1B-Chat-v1.0 \
    --dataset_name wikitext \
    --dataset_config_name wikitext-2-raw-v1 \
    --per_device_train_batch_size 1 \
    --per_device_eval_batch_size 1 \
    --do_train \
    --do_eval \
    --tp_size 2 \
    --output_dir /tmp/test-clm

**unexpected behavior:**
runtimeerror: aten._foreach_norm_Scalar: got mixed torch.tensor and DTensor， need to convert all torch.tensor to DTensor before calling distributed operators. 

### Expected behavior

autoTP training

### PR DESCRIPTION
**Pull request by kmehant**
# What does this PR do?
`torch.nn.utils.clip_grad_norm_` does not support heterogenous set of parameters having a mix of DTensors and Tensors. This PR allows for gradient clipping by distributing non-sharded modules that are not involved in TP. We replicate all such modules across the device mesh. 

The PR also adds new parallel style `ReplicateParallel` so that the existing TP APIs can be used as is for this module replication operation. We could think of contributing this back to PyTorch if it makes sense (cc: @kwen2501) otherwise we can maintain it in transformers.

**⭐ Note : We would rebase this PR once https://github.com/huggingface/transformers/pull/34194 is merged some of the workflow changes that you see here would disappear once the PR is merged.**

fixes: #36296

# Concerns

### Concern 1
When we do two TP runs with gradient clipping with exact same training settings we dont reproduce exact loss parity between the runs though both the runs converge eventually. I am worried if `Replicate` sharding has something to do here.
Screenshot 2025-02-11 at 6 51 37 PM

### Concern 2
Grad norms are not same on each rank, I would assume in TP training the grad norms should come out to be same across the ranks however thats not the case

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@ArthurZucker @muellerzr and @SunMarc
@kwen2501 from PyTorch

### CODE REVIEW

**Comment by SunMarc** [comment]
Can you have a look @kwen2501 ?

**Comment by kwen2501** [comment]
@weifengpy @mori360 do you mind having a look at the two concerns here? Thanks!

**Comment by weifengpy** [comment]
> @weifengpy @mori360 do you mind having a look at the two concerns here? Thanks!

> does not support heterogenous set of parameters having a mix of DTensors and Tensors

`implicit_replication` is invented to mix DTensor with plain tensors. Maybe it's cleaner here. 

```
from torch.distributed._tensor.experimental import implicit_replication
with implicit_replication():
   # call gradient clipping here
```

code pointer: https://github.com/pytorch/pytorch/blob/8b818ab58f635f999de2c8a5bf8e6c01d0c122ed/test/distributed/tensor/parallel/test_tp_examples.py#L262-L264

**Comment by kmehant** [comment]
@weifengpy Do you recommend to use `implicit_replication` instead?

File: src/transformers/models/granite/configuration_granite.py [hunk 1]
```diff
@@ -117,10 +117,14 @@ class GraniteConfig(PretrainedConfig):
...
-        "layers.*.self_attn.o_proj": "rowwise",
+        "layers.*.self_attn.o_proj": "rowwise_output_dtensor",
         "layers.*.mlp.gate_proj": "colwise",
         "layers.*.mlp.up_proj": "colwise",
-        "layers.*.mlp.down_proj": "rowwise",
+        "layers.*.mlp.down_proj": "rowwise_output_dtensor",
+        "embed_tokens": "replicateparallel_output_dtensor",
+        "layers.*.post_attention_layernorm": "replicateparallel_output_dtensor",
+        "layers.*.input_layernorm": "replicateparallel_output_dtensor",
+        "norm": "replicateparallel_output_dtensor",
```
Comment (kwen2501) [inline src/transformers/models/granite/configuration_granite.py:127]: Thanks for extending the configs here.
I wonder if some of these settings would be more interesting to training than to inference? 
(On the other hand, I don't know much about HF's user profile -- training more or inference more?)
If some of the settings are specific to training, is it possible to separate them out? Or, shall we make the config somehow customizable at run time?

File: src/transformers/pytorch_utils.py [hunk 2]
```diff
@@ -344,8 +351,82 @@ def isin_mps_friendly(elements: torch.Tensor, test_elements: torch.Tensor | int)
         return torch.isin(elements, test_elements)
 
 
-# TODO need to add the __repr__ that shows that it is a colwise parallel
-# See https://github.com/pytorch/pytorch/issues/145726
+class ReplicateParallel(ParallelStyle):
+    """
+    Replicate a nn.Module.
+    Users can compose it together with other parallel styles like RowwiseParallel to achieve a fully distributed model.
+    Fully distributed model is needed for gradient clipping.
```
Comment (kwen2501) [inline src/transformers/pytorch_utils.py:358]: @weifengpy @wz337 @tianyu-l 
I wonder if there is anything we can do on DTensor side so that users don't have to annotate the entire model to perform gradient clipping?

File: src/transformers/pytorch_utils.py [hunk 3]
```diff
@@ -344,8 +351,82 @@ def isin_mps_friendly(elements: torch.Tensor, test_elements: torch.Tensor | int)
         return torch.isin(elements, test_elements)
 
 
-# TODO need to add the __repr__ that shows that it is a colwise parallel
-# See https://github.com/pytorch/pytorch/issues/145726
```
Comment (kwen2501) [inline src/transformers/pytorch_utils.py:348]: nit: keep this TODO?

**Review by kwen2501 — COMMENTED** [review]
Thanks for fixing gradient clipping with TP. 
Functionality wise, the code change looks reasonable. 
I am consulting with my colleagues to see if there is a must to explicitly annotate non-sharded modules as Replicated.

**Review by ArthurZucker — COMMENTED** [review]
Very nice, waiting on @kwen2501 's feed back, but make sure to rebase since we just merged: #36335 !

**Comment by ArthurZucker** [comment]
@kmehant we finish the refactoring if you still want to work on this!

**Comment by kmehant** [comment]
https://github.com/huggingface/transformers/pull/36132#issuecomment-2713283865

Thanks for the update, I will soon rebase the PR by EoD. Thanks

**Comment by kmehant** [comment]
@ArthurZucker I have rebased my PR, thanks

File: src/transformers/pytorch_utils.py (hunk 2 above)
Comment (kmehant) [reply src/transformers/pytorch_utils.py:358]: @weifengpy  @kwen2501 Should we be using `implicit_replication()` as an alternative?

File: src/transformers/pytorch_utils.py (hunk 3 above)
Comment (kmehant) [reply src/transformers/pytorch_utils.py]: Fixed thanks

**Comment by kwen2501** [comment]
Yeah, per @weifengpy 's comment, I think `implicit_replication()` is preferred over creating new strategies. You can limit its usage to minimal possible if there is risk concern -- like, only when `clip_grad_norm_` is used.

**Review by ArthurZucker — APPROVED** [review]
Thanks! 
I am not familiar with `implicit_replication` but if it is recommended by @kwen2501 happy to use! I suppose it requires a specific torch version check to use no? 

If so, using our own replicate would be a bit better AFAIK for broader support (starting 2.3 vs only version where implicit replication is defined

File: src/transformers/trainer.py [hunk 4]
```diff
@@ -234,6 +234,7 @@
         AutocastKwargs,
         DistributedDataParallelKwargs,
         DistributedType,
+        TorchTensorParallelPlugin,
```
Comment (ArthurZucker) [inline src/transformers/trainer.py:237]: not seing this used!

**Review by SunMarc — APPROVED** [review]
much needed !
